Add deprecation logging message for 'fuzzy' query

cbuescher opened the pull request:
The `fuzzy` query is deprecated from 5.0 on. Similar to IndicesQueryBuilder we should log a deprecation warning whenever this query is used.

Relates to #15760

cbuescher committed:
Add deprecation logging message for 'fuzzy' query

This query is deprecated from 5.0 on. Similar to IndicesQueryBuilder we should
log a deprecation warning whenever this query is used.

Relates to #15760

cbuescher commented:
This should also go to 5.x and maybe 5.0, since the query is deprecated from 5.0 on.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FuzzyQueryBuilder.java:
```diff
@@ -151,6 +155,7 @@ public FuzzyQueryBuilder(String fieldName, boolean value) {
      * @param value The value of the term
      */
     public FuzzyQueryBuilder(String fieldName, Object value) {
+        DEPRECATION_LOGGER.deprecated("{} query is deprecated. Instead use the [match] query with fuzziness parameter", NAME);
```
add a test for this? If I remember correctly the deprecation logger is hooked into the response header warnings, hence testable from the outside without looking at logs.

nik9000 replied:
> Do you know if there's a way of checking this in the regular REST (yaml) tests?

You can look at the code generated by the docs tests and do what it does. You should add a `feature` section to the test so the clients don't choke on it.

Personally I'm fine with just asserting it in the docs tests. The build still fails if we remove it or change the message in some crazy way and that is all I'm looking for.

cbuescher replied:
@javanna I looked into adding rest tests that check the warnings like @nik9000 suggested, I could add a test similar to the one that the docs-tests generate but there seems to be no other rest test for "fuzzy" at the moment and it feels strange to introduce tests for a deprecated query. As mentioned above, the docs tests already check the warning headers. Are you okay with leaving it like this?

javanna replied:
I am on the fence. I think we should have a test for this, but I understand your objections. I am not sure we should rely on docs tests to verify this though. Docs tests are there to ensure that docs are correct, and we shouldn't be tempted to use them to verify that the code is correct. Ideally we should have a failing test besides docs if we remove the deprecation warning. Maybe add a unit test to fuzzy query builder tests that verifies that the deprecation log goes somewhere (in the ThreadContext). Is that feasible?

nik9000 replied:
I'm quite happy to use docs tests to verify that the code is correct in lots of cases, mostly the cases where you are going to document the feature anyway and there isn't a lot of value having all the clients test it. Then you get docs and you get tests. I don't think you should add useless docs just because they add tests, but I think we could do better to document more things are if that lets us skip adding yaml tests for them then great, incentive to do the right thing.

I like the yaml tests (and junit tests with rest client) to do more complex edge case stuff that you aren't going to document.

javanna reviewed (changes requested):
left one comment

cbuescher commented:
@javanna I added a test to the FuzzyQueryBuilder unit test, but I'm not too sure if this is the best way to test this. If you have any other idea how this can be done in our unit tests please let me know.

javanna commented:
>  I added a test to the FuzzyQueryBuilder unit test, but I'm not too sure if this is the best way to test this. If you have any other idea how this can be done in our unit tests please let me know.

I would have done it exactly that way, this is how I envision we will test deprecations warnings from now on. What is wrong with that? :) Much better than having to write an integration test for it.

cbuescher commented:
Thanks @javanna , I just wasn't sure setting the ThreadLocal on the logger in the test is the right way to go, since it might be used by other tests elsewhere (thats why I added those comments). So I assume this is okay then, should I merge this?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTests.java:
```diff
@@ -181,4 +187,16 @@ public void testParseFailsWithMultipleFields() throws IOException {
     }
+
+    public void testDeprecationLoggingHeaders() throws IOException {
+        try (ThreadContext threadContext = new ThreadContext(Settings.EMPTY)) {
```
Maybe add the threadContext creation in the base class? and set it / remove it in the before /after test ?

cbuescher replied:
I am on the fence. If other tests are executed concurrently, then adding the ThreadContext to the DeprecationLogger in the base class will mean that it potentially receives deprecation warnings from elsewhere. I was hoping to make this as local as possible so I'm almost sure that we test the following contructor call in isolation. So I'd rather add the context here, and even remove it from the Deprecation logger after this test is done. I'll push an update along that line.

javanna replied:
No tests will ever be executed concurrently. Removing the thread context is enough, I suspect that this is how we will test deprecation warnings once we remove strict parsing, which we currently use only for testing.

cbuescher commented:
@javanna I decided to go the route suggested by @nik9000 and add the deprecation warning checks to all query test (and the QueryParseContextTests). I pushed a separate commit since this affects tests other than for `fuzzy`, so the scope got a big bigger than the original intention of this PR. I can also open a separate PR for this if you like. Basically we add a new ThreadContext to the DeprecationLogger in @Before and check that it is empty in @After in AbstractQueryTestCase. Each test that triggers some warning headers needs to check them explicitely (which will also reset the logger to a new ThreadContext if successful). I hope this is along the lines you envisioned it.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTests.java:
```diff
@@ -348,9 +348,14 @@ public void testFromJsonEmptyQueryBody() throws IOException {
+        // we should have deprecation warning headers regardless of throwing and exception
+        checkWarningHeaders("query malformed, empty clause found at [1:27]",
+                "query malformed, empty clause found at [1:46]",
+                "query malformed, empty clause found at [1:100]");
```
nice! comment should be "throwing an exception" I guess.

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTests.java:
```diff
@@ -38,6 +39,14 @@
+     * all tests create deprecation warnings through the builders constructor, those are
+     * checked once here after test it run
+     */
+    @After void checkWarningHeaders() {
```
missing line break after annotation?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTests.java:
```diff
@@ -38,6 +39,14 @@
 
+    /**
+     * all tests create deprecation warnings through the builders constructor, those are
+     * checked once here after test it run
```
what do you mean by "the builders constructor" ? Should it be "after each test is run" ?

javanna commented on core/src/test/java/org/elasticsearch/index/query/IndicesQueryBuilderTests.java:
```diff
@@ -22,11 +22,20 @@
 
+    /**
+     * all tests create deprecation warnings through the builders constructor, those are
+     * checked once here after test it run
```
same comments as above in FuzzyQueryBuilderTests

javanna commented on core/src/test/java/org/elasticsearch/index/query/QueryParseContextTests.java:
```diff
@@ -45,6 +52,19 @@ public static void init() {
+
+    @After
+    public void teardown() throws IOException {
+        this.threadContext.close();
```
shouldn't we do DeprecationLogger.removeThreadContext first?

javanna commented on core/src/test/java/org/elasticsearch/index/query/QueryParseContextTests.java:
```diff
@@ -89,6 +109,10 @@ public void testParseInnerQueryBuilderEmptyBody() throws IOException {
             QueryParseContext context = new QueryParseContext(indicesQueriesRegistry, parser, ParseFieldMatcher.EMPTY);
             Optional<QueryBuilder> emptyQuery = context.parseInnerQueryBuilder();
             assertFalse(emptyQuery.isPresent());
+            // regardless of strict mode, we should get warnings in the response headers
```
remove this comment? strict mode will be gone soon, one less edit needed later :)

javanna commented on modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/TemplateQueryBuilderTests.java:
```diff
@@ -56,6 +57,14 @@
 
+    /**
+     * all tests create deprecation warnings through the builders constructor, those are
+     * checked once here after test it run
```
same comments as above

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -1026,6 +1052,18 @@ protected Query rewrite(Query query) throws IOException {
+            assertThat(warnings, hasItem(equalTo(msg)));
+        }
+        // "clear" current warning headers by setting a new ThreadContext
+        DeprecationLogger.removeThreadContext(this.threadContext);
```
I think that we are leaking a thread local here? We should close the current threadContext before overriding it.

cbuescher replied:
yes

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -153,6 +159,10 @@
+    /**
+     * used to check warning headers of the deprecation logger
+     */
+    protected ThreadContext threadContext;
```
does this need to be protected? I think it is safer if it is private and the only way to access it is via checkWarningHeaders

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -1026,6 +1052,18 @@ protected Query rewrite(Query query) throws IOException {
+        // "clear" current warning headers by setting a new ThreadContext
+        DeprecationLogger.removeThreadContext(this.threadContext);
+        this.threadContext = new ThreadContext(Settings.EMPTY);
+        DeprecationLogger.setThreadContext(this.threadContext);
```
It would be nice to find another way to do this other than replacing the thread context. I had a look but didn't find other ways though :) I guess you have tried as well.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java:
```diff
@@ -207,6 +217,22 @@ public void beforeTest() throws IOException {
+    @After
+    public void teardown() throws IOException {
+        final List<String> warnings = threadContext.getResponseHeaders().get(DeprecationLogger.DEPRECATION_HEADER);
+        if (warnings != null) {
```
do we expect null or empty when there are no warnings? seems like both are ok but two invariants are not needed and only one should be the right one?

cbuescher replied:
I think `null` is what you get back from the response headers map if there are no "Warning" entries.

javanna reviewed (changes requested):
I left a few comments but I like the latest changes. This is totally going the direction I wanted it to go, thanks! You are already making changes that I had made when I first started working on #19552. With this new way of checking deprecation warnings, we almost don't need strict parsing anymore in our tests.

cbuescher commented:
@javanna thanks, I rebased and added a commit trying to address your comments.

javanna commented on core/src/test/java/org/elasticsearch/index/query/FuzzyQueryBuilderTests.java:
```diff
@@ -107,7 +113,12 @@ public void testUnsupportedFuzzinessForStringType() throws IOException {
     public void testToQueryWithStringField() throws IOException {
-        assumeTrue("test runs only when at least a type is registered", getCurrentTypes().length > 0);
+        try {
+            assumeTrue("test runs only when at least a type is registered", getCurrentTypes().length > 0);
```
shall we move to using `@ConditionalIgnore` here. If I understood correctly in that case before and after wouldn't be executed for ignored methods, hence you wouldn't have to keep track of whether the test ran yourself

cbuescher replied:
I didn't know that one, great suggestion. I will try if this works.

cbuescher replied:
Unfortunately @ConditionalIgnore doesn't seem to be part of junits standard set of annotations yet (seems to be targeted for junit 5). Rather than build our own or import one of the custom solutions I found, I'd keep the current solution. This whole test class is going away with the deprecated query builder anyway soon (will stay on the 5.x branch for it's lifetime however).

javanna reviewed (changes requested):
left one comment on using assumptions, LGTM otherwise

javanna reviewed (approved):
LGTM

cbuescher commented:
@javanna thanks, I'll rebase an squash a bit, then merge this to master and 5.x

cbuescher committed:
Tests: Add warning header checks to QueryBuilder tests and QueryParseContextTests

This adds checks for expected warning headers to the query builder test
infrastructure. Tests that are adding deprecation warnings to the response
headers need to check those, otherwise the abstract base class for the test
class will complain at teardown.

cbuescher commented:
Merged to 5.x with 56cffc68c93616386e70304cb0e542ea1c05e117 and 5806bb6b230ab410f9ae0d7013ebdf60e95ba0d6
